cmd/dep/root_analyzer.go

```diff
@@ -139,6 +140,26 @@ func (a *rootAnalyzer) FinalizeRootManifestAndLock(m *dep.Manifest, l *dep.Lock)
 			delete(m.Constraints, pr)
 		}
 	}
+	// Pick the direct dependencies from the solution lock and add to manifest.
+	// This is done to fill up the manifest constraints with the dependencies
+	// solved over the network.
+	for _, y := range l.Projects() {
+		var f *fb.ConstraintFeedback
+		pr := y.Ident().ProjectRoot
+		if _, ok := a.directDeps[string(pr)]; ok {
+			pp := getProjectPropertiesFromVersion(y.Version())
+			if pp.Constraint != nil {
+				m.Constraints[pr] = pp
+				pc := gps.ProjectConstraint{Ident: y.Ident(), Constraint: pp.Constraint}
+				f = fb.NewConstraintFeedback(pc, fb.DepTypeDirect)
+			} else {
+				f = fb.NewLockedProjectFeedback(y, fb.DepTypeDirect)
+			}
+		} else {
+			f = fb.NewLockedProjectFeedback(y, fb.DepTypeTransitive)
+		}
+		f.LogFeedback(a.ctx.Err)
+	}
```

Comment:
I remember we used to keep a copy of lock given to the solver and use the same to identify new constraints and projects, by comparing with new lock, at one point of time. I still have my dep fork at that point of time, [copyLock](https://github.com/darkowlzz/dep/blob/master/cmd/dep/init.go#L162) 😊

So, yeah, a unified `FinalizeManifestAndLock` is what we need 🙌